libtsuba/include/tsuba/ReadGroup.h

```diff
@@ -0,0 +1,68 @@
+#ifndef KATANA_LIBTSUBA_TSUBA_READGROUP_H_
+#define KATANA_LIBTSUBA_TSUBA_READGROUP_H_
+
+#include <future>
+#include <list>
+#include <memory>
+
+#include "katana/Result.h"
+#include "tsuba/AsyncOpGroup.h"
+
+namespace arrow {
+class Table;
+};
+
+namespace tsuba {
+
+/// Track multiple, outstanding async writes and provide a mechanism to ensure
+/// that they have all completed
+class ReadGroup {
+public:
+  static katana::Result<std::unique_ptr<ReadGroup>> Make();
+
+  /// Wait until all operations this descriptor knows about have completed
+  katana::Result<void> Finish();
+
+  /// Add future to the list of futures this ReadGroup will wait for, note
+  /// the file name for debugging. `on_complete` is guaranteed to be invoked
+  /// after `on_complete`'s passed by previous calls to AddOp
+  void AddOp(
+      std::future<katana::Result<void>> future, std::string file,
+      const std::function<katana::Result<void>()>& on_complete);
+
+  /// same as AddOp, but the future may return a data type which can then be
+  /// consumed by on_complete
+  void AddReturnsOp(
+      std::future<katana::Result<std::shared_ptr<arrow::Table>>> future,
+      const std::string& file,
+      const std::function<katana::Result<void>(std::shared_ptr<arrow::Table>)>&
+          on_complete) {
+    // n.b., make shared instead of unique because move capture below prevents
+    // passing generic_complete_fn as a std::function
+    auto ret_val = std::make_shared<std::shared_ptr<arrow::Table>>();
+    auto new_future = std::async(
```

Comment:
It's worse; here i'm making a pointer to a pointer as a piece of shared memory that my two lambdas below can use to communicate.

Reply:
Yeah, that does take you down the rabbit hole.  I'd at least add that as a comment.

Reply:
I added a comment, and I actually templated the function as well (was my goal so that this works for return types besides `std::shared_ptr<arrow::Table>`).  I think the templating also makes the intent clearer since you don't see the confusing, duplicated `shared` on the line.